ENH: add optional ddof param to DataFrame.cov()

edhalter opened the pull request:
- [ ] closes #xxxx (searched, didn't see any issues on this topic)
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

Noticed that the .cov() methods on DataFrame and Series assume the user wants to calculate sample variance and normalize by N-1. In contrast, the .var() and .std() methods on DataFrame, Series, DataFrameGroupBy, and SeriesGroupBy take a ddof parameter. Its default value is 1, calculating the sample std/var, but the user can specify ddof=0 to calculate the population std/var.

Simply added a ddof param (default value 1) to the .cov() methods. The value is then passed into other method calls as required. Did have to modify 1 bit of Cython code in algos.pyx, the nancorr() function.

edhalter committed:
ENH: add optional ddof param to .cov() (GH16227)

jreback commented on pandas/core/frame.py:
```diff
@@ -4968,15 +4968,20 @@ def corr(self, method='pearson', min_periods=1):
             to have a valid result.
+        ddof : int, default 1
+            Delta Degrees of Freedom.  The divisor used in calculations
+            is ``N - ddof``, where ``N`` represents the number of elements.
```
versionadded tag

jreback commented on pandas/core/series.py:
```diff
@@ -1374,15 +1374,20 @@ def corr(self, other, method='pearson', min_periods=None):
+        ddof : int, default 1
+            Delta Degrees of Freedom.  The divisor used in calculations
+            is ``N - ddof``, where ``N`` represents the number of elements.
 
```
same

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -27,6 +27,9 @@ New features
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- The ``.cov()`` methods for DataFrame and Series now accept the parameter
```
add this PR number here

jreback replied:
accept -> accepts

jreback reviewed (approved):
lgtm.

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -4968,15 +4968,20 @@ def corr(self, method='pearson', min_periods=1):
-        Compute pairwise covariance of columns, excluding NA/null values
+        Compute pairwise covariance of columns, excluding NA/null values.
+        
+        Normalized by N-1 by default. This can be changed using the ddof argument"
```
I don't think this line is needed (as it's have it in the Notes section)

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -4985,7 +4990,8 @@ def cov(self, min_periods=None):
         -----
         `y` contains the covariance matrix of the DataFrame's time series.
-        The covariance is normalized by N-1 (unbiased estimator).
+        The covariance is normalized by N-ddof -- for the default ddof value of
```
Double backticks around `N-ddof` and `ddof`.

TomAugspurger commented on pandas/core/series.py:
```diff
@@ -1374,15 +1374,20 @@ def corr(self, other, method='pearson', min_periods=None):
         """
         Compute covariance with Series, excluding missing values
+        
+        Normalized by N-1 by default. This can be changed using the ddof argument"
```
Can probably make this a new Notes section, and combine with the note that's currently in the Returns section.

jreback commented:
can you rebase and update?

edhalter commented:
Made changes to docstrings, etc. based on comments. Rebased and updated.

TomAugspurger commented:
@edhalter did you push your changes? There seems to be a conflict in the whatsnew file. You may need to force push to your branch.

edhalter commented:
Already pushed -- the conflicting file has my most recent changes. Someone just needs to edit it and append my new line to the other new lines under "Other Enhancements".

jreback commented:
@edhalter pls rebase on master. this will present a conflict which you fix, then push.

edhalter commented:
Rebased on master.

jreback commented on pandas/core/frame.py:
```diff
@@ -5005,15 +5005,20 @@ def corr(self, method='pearson', min_periods=1):
             Minimum number of observations required per pair of columns
             to have a valid result.
+        ddof : int, default 1
+            .. versionadded:: 0.21.0
```
move the versionadded tag below the expl.

jreback commented on pandas/core/frame.py:
```diff
@@ -5022,7 +5027,8 @@ def cov(self, min_periods=None):
         -----
         `y` contains the covariance matrix of the DataFrame's time series.
-        The covariance is normalized by N-1 (unbiased estimator).
+        The covariance is normalized by ``N-ddof`` -- for the default ``ddof``
```
just start a new sentence, I don't think '--' works here.

jreback commented on pandas/core/series.py:
```diff
@@ -1408,18 +1408,26 @@ def cov(self, other, min_periods=None):
         min_periods : int, optional
             Minimum number of observations needed to have a valid result
+        ddof : int, default 1
+            .. versionadded:: 0.21.0
```
same as above

jreback commented on pandas/core/series.py:
```diff
@@ -1408,18 +1408,26 @@ def cov(self, other, min_periods=None):
-        Normalized by N-1 (unbiased estimator).
+        Notes
+        -----
+        The returned covariance is normalized by ``N-ddof`` -- for the default
```
same as above

jreback commented on pandas/_libs/algos.pyx:
```diff
@@ -325,7 +330,7 @@ def nancorr(ndarray[float64_t, ndim=2] mat, bint cov=0, minp=None):
                             sumxx += vx * vx
                             sumyy += vy * vy
 
-                    divisor = (nobs - 1.0) if cov else sqrt(sumxx * sumyy)
```
what about ddofv >= nobs case? can you add a test for this. looks like divisor == 0 is handled, but what about ddofv > nobs?

edhalter committed:
raise on ddof < 0, warn on N - ddof <= 0

edhalter committed:
lint fixes

edhalter commented:
Added some checks in DataFrame.cov() and the nancov() func called by Series.cov():

- sanity check on input ddof
- warn when N - ddof <= 0

For the latter, alternatively, could throw an Exception, or just silently return nan. The warning seemed better -- that's what numpy's cov() does:

`warnings.warn("Degrees of freedom <= 0 for slice", RuntimeWarning)`

However, the Cython nancorr() func called by DataFrame.cov() does just silently return nan.

jreback commented:
> warnings.warn("Degrees of freedom <= 0 for slice", RuntimeWarning)

can you change UserWarning -> RuntimeWarning to be consistent

> However, the Cython nancorr() func called by DataFrame.cov() does just silently return nan.

can you change (and test) this as well?

jreback commented:
if you can rebase and update

jreback commented:
can you rebase

edhalter committed:
change warning types; add warning to nancorr()

edhalter committed:
clean up frame's test_cov() and add extra checks

pep8speaks commented:
Hello @edhalter! Thanks for updating the PR.







 - In the file [`pandas/tests/frame/test_analytics.py`](https://github.com/pandas-dev/pandas/blob/7cb4b0b62b942fd39d933e99df75d0c37f18d546/pandas/tests/frame/test_analytics.py), following are the PEP8 issues :

> [Line 137:30](https://github.com/pandas-dev/pandas/blob/7cb4b0b62b942fd39d933e99df75d0c37f18d546/pandas/tests/frame/test_analytics.py#L137): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 139:80](https://github.com/pandas-dev/pandas/blob/7cb4b0b62b942fd39d933e99df75d0c37f18d546/pandas/tests/frame/test_analytics.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 146:80](https://github.com/pandas-dev/pandas/blob/7cb4b0b62b942fd39d933e99df75d0c37f18d546/pandas/tests/frame/test_analytics.py#L146): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

edhalter committed:
lint fixes

edhalter commented:
Returning to this.... Changed warning type to RuntimeWarning, and added a single warning that will throw in cython from nancorr() if at least 1 result in the matrix is set to NaN due to N - ddof <= 0.

Manual tests looked good. Also added some more automated tests in DataFrame's test_cov().

jreback commented on pandas/core/frame.py:
```diff
@@ -5307,23 +5312,33 @@ def cov(self, min_periods=None):
         if notna(mat).all():
+            fill_nan = False
             if min_periods is not None and min_periods > len(mat):
+                fill_nan = True
```
can you explain whats going on here

jreback commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -129,31 +129,55 @@ def test_corr_cov_independent_index_column(self):
+
+        # compare each number in DataFrame.cov() matrix
+        # to corresponding Series.cov() number
+        for ddof in [0, 1]:
```
can you do this

jreback reviewed (changes requested):
pls rebase and update according to comments

jreback commented:
closing as stale, but if you'd like to continue work, pls rebase and ping.
